### PR TITLE
server: avoid reading from etcd if possible

### PR DESCRIPTION
**Pull request by disksing**
## What have you changed? (required)
When leader node of the cluster is down or isolated, all reads will fail after timeout. We can see many errors are reported in logs:
```
2018/07/27 10:01:51.709 log.go:82: [warning] etcdserver: [read-only range request "key:\"/pd/6582711075678387698/leader\" " took too long (10.000606988s) to execute]
2018/07/27 10:01:51.709 etcd_kv.go:116: [error] load from etcd error: context deadline exceeded
2018/07/27 10:01:51.709 etcd_kv.go:119: [warning] kv gets too slow: key /pd/6582711075678387698/leader cost 10.001256856s err context deadline exceeded
2018/07/27 10:01:51.709 grpc_service.go:61: [error] failed to load pd leader: context deadline exceeded
2018/07/27 10:01:51.723 log.go:82: [warning] etcdserver: [read-only range request "key:\"/pd/6582711075678387698/leader\" " took too long (9.999997133s) to execute]
2018/07/27 10:01:51.723 etcd_kv.go:116: [error] load from etcd error: context deadline exceeded
2018/07/27 10:01:51.723 etcd_kv.go:119: [warning] kv gets too slow: key /pd/6582711075678387698/leader cost 10.00029717s err context deadline exceeded
2018/07/27 10:01:51.723 grpc_service.go:61: [error] failed to load pd leader: context deadline exceeded
2018/07/27 10:01:51.901 etcd_kv.go:116: [error] load from etcd error: context deadline exceeded
```
and
```
2018/07/27 09:39:30.843 log.go:82: [warning] etcdserver: [read-only range request "key:\"/pd/6582704709390457557/member/1162990608748153514/leader_priority\" " took too long (9.005324225s) to execute]
2018/07/27 09:39:30.849 log.go:82: [warning] etcdserver: [read-only range request "key:\"/pd/6582704709390457557/member/1162990608748153514/leader_priority\" " took too long (4.22620823s) to execute]
2018/07/27 09:39:30.850 etcd_kv.go:119: [warning] kv gets too slow: key /pd/6582704709390457557/member/1162990608748153514/leader_priority cost 9.012000983s err <nil>
2018/07/27 09:39:30.851 log.go:82: [warning] etcdserver: [read-only range request "key:\"/pd/6582704709390457557/member/1162990608748153514/leader_priority\" " took too long (913.222306ms) to execute]
2018/07/27 09:39:30.851 log.go:82: [warning] etcdserver: [read-only range request "key:\"/pd/6582704709390457557/member/1162990608748153514/leader_priority\" " took too long (914.525254ms) to execute]
2018/07/27 09:39:30.851 log.go:82: [warning] etcdserver: [read-only range request "key:\"/pd/6582704709390457557/member/1162990608748153514/leader_priority\" " took too long (914.644281ms) to execute]
```

What's worse, the pending reads will affect leader election speed when the node is back.

As a workr, we need to avoid reading from etcd if it is possible, especially when etcd cluster has no leader node.

The changes including:
1. Skip loading leader priority in gRPC method `GetMember()`, only load them in HTTP api '/members' handler
2. Before loading leader or leader priorities from etcd, check if the etcd leader is elected. This is done by check if `etcd.Server.Lead()` is non-zero. Be aware that it is not always up to date, but it can prevent etcd read be blocked in most cases.
3. Cache current leader of PD cluster in `Server`. For gRPC and API calls, we don't have to read etcd anymore.

## What are the type of the changes (required)?
- Improvement (non-breaking change which is an improvement to an existing feature)

## How has this PR been tested (required)?
- make dev
- manual test

### CODE REVIEW

File: server/leader.go [hunk 1]
```diff
@@ -36,16 +35,33 @@ var (
...
+	return s.GetLeader().GetMemberId()
+}
+
+// GetLeader returns current leader of pd cluster.
+func (s *Server) GetLeader() *pdpb.Member {
+	leader := s.leader.Load()
+	if leader == nil {
+		return nil
 	}
+	member := leader.(*pdpb.Member)
```
Comment (gregwebs) [inline server/leader.go]: Can we do the type assertion in a loading function?

File: server/api/server_test.go [hunk 2]
```diff
@@ -110,16 +111,28 @@ func mustNewCluster(c *C, num int) ([]*server.Config, []*server.Server, cleanUpF
...
+		var leader *pdpb.Member
 		for _, svr := range svrs {
-			if svr.IsLeader() {
-				return svr
+			l := svr.GetLeader()
+			if l == nil {
+				return false
+			}
+			// All servers' GetLeader should return the same leader.
+			if l == nil || (leader != nil && l.GetMemberId() != leader.GetMemberId()) {
```
Comment (rleungx) [inline server/api/server_test.go]: duplicated?

File: server/server.go [hunk 3]
```diff
@@ -72,7 +72,8 @@ type Server struct {
 	id          uint64 // etcd server id.
 	clusterID   uint64 // pd cluster id.
 	rootPath    string
-	leaderValue string // leader value saved in etcd leader key.  Every write will use this to check leader validation.
+	member      *pdpb.Member // current PD's info.
+	leaderValue string       // leader value saved in etcd leader key.  Every write will use this to check leader validation.
```
Comment (rleungx) [inline server/server.go]: an extra space before `Every`

File: server/leader.go [hunk 4]
```diff
@@ -165,26 +184,16 @@ func getLeader(c *clientv3.Client, leaderPath string) (*pdpb.Member, error) {
...
+func (s *Server) GetEtcdLeader() uint64 {
+	return s.etcd.Server.Lead()
 }
 
 func (s *Server) isSameLeader(leader *pdpb.Member) bool {
 	return leader.GetMemberId() == s.ID()
 }
 
-func (s *Server) marshalLeader() string {
+func (s *Server) memberInfo() (member *pdpb.Member, marshalStr string) {
```
Comment (siddontang) [inline server/leader.go]: leader may be better for the return name?

File: server/leader.go [hunk 5]
```diff
@@ -30,22 +29,36 @@ import (
...
+	return member
+}
+
+func (s *Server) enableLeader() {
+	s.leader.Store(s.member)
+}
 
-	atomic.StoreInt64(&s.isLeader, value)
+func (s *Server) disableLeader() {
+	s.leader.Store(&pdpb.Member{})
```
Comment (siddontang) [inline server/leader.go]: why not setting nil directly?

File: server/leader.go [hunk 6]
```diff
@@ -294,7 +303,10 @@ func (s *Server) campaignLeader() error {
 	}
 }
 
-func (s *Server) watchLeader() {
+func (s *Server) watchLeader(leader *pdpb.Member) {
+	s.leader.Store(leader)
+	defer s.leader.Store(&pdpb.Member{})
```
Comment (siddontang) [inline server/leader.go]: use nil?

File: server/server.go [hunk 7]
```diff
@@ -198,7 +199,7 @@ func (s *Server) startServer() error {
 	metadataGauge.WithLabelValues(fmt.Sprintf("cluster%d", s.clusterID)).Set(0)
 
 	s.rootPath = path.Join(pdRootPath, strconv.FormatUint(s.clusterID, 10))
-	s.leaderValue = s.marshalLeader()
+	s.member, s.leaderValue = s.memberInfo()
```
Comment (siddontang) [inline server/server.go]: Is `s.member` for leader or for current PD? seems `memberInfo` returns the leader?

Confusing...

File: server/leader.go (hunk 5 above)
Comment (disksing) [reply server/leader.go]: It will panic if you store nil. See https://godoc.org/sync/atomic#Value.Store

File: server/server.go (hunk 7 above)
Comment (disksing) [reply server/server.go]: Current PD's member information. It will be saved to leader key if the PD successfully elected as leader.

Comment (siddontang) [reply server/server.go]: em, I think using memberValue may be more sensible. For me, leaderValue seems the value from the leader and the leader may be another node...

Comment (disksing) [reply server/server.go]: good point.

**Pull request merged**